libvips/resample/reduceh.cpp

```diff
@@ -435,22 +438,61 @@ vips_reduceh_build( VipsObject *object )
 	VipsResample *resample = VIPS_RESAMPLE( object );
 	VipsReduceh *reduceh = (VipsReduceh *) object;
 	VipsImage **t = (VipsImage **) 
-		vips_object_local_array( object, 2 );
+		vips_object_local_array( object, 3 );
 
 	VipsImage *in;
-	double width, extra_pixels;
+	int width;
+	int int_hshrink;
+	double extra_pixels;
 
 	if( VIPS_OBJECT_CLASS( vips_reduceh_parent_class )->build( object ) )
 		return( -1 );
 
 	in = resample->in; 
 
-	if( reduceh->hshrink < 1 ) { 
+	if( reduceh->hshrink < 1.0 ) { 
 		vips_error( object_class->nickname, 
-			"%s", _( "reduce factors should be >= 1" ) );
+			"%s", _( "reduce factor should be >= 1.0" ) );
 		return( -1 );
 	}
 
+	/* Output size. We need to always round to nearest, so round(), not
+	 * rint().
+	 */
+	width = VIPS_ROUND_UINT(
+		(double) in->Xsize / reduceh->hshrink );
+
+	/* How many pixels we are inventing in the input, -ve for
+	 * discarding.
+	 */
+	extra_pixels = width * reduceh->hshrink - in->Xsize;
+
+	if( reduceh->gap != 0.0 &&
+		reduceh->kernel != VIPS_KERNEL_NEAREST ) {
+		if( reduceh->gap < 1.0 ) {
+			vips_error( object_class->nickname,
+				"%s", _( "reduce gap should be >= 1.0" ) );
+			return( -1 );
+		}
+
+		/* The int part of our reduce.
+		 */
+		int_hshrink = VIPS_MAX( 1,
+			VIPS_FLOOR( (double) in->Xsize / width / reduceh->gap ) );
+
+		if( int_hshrink > 1 ) {
+			g_info( "shrinkh by %d", int_hshrink );
+			if( vips_shrinkh( in, &t[0], int_hshrink,
+				"ceil", TRUE,
+				NULL ) )
+				return( -1 );
+			in = t[0];
+
+			reduceh->hshrink /= int_hshrink;
+			extra_pixels /= int_hshrink;
+		}
+	}
+
 	if( reduceh->hshrink == 1 ) 
```

Comment:
Maybe `1.0` here too (I realize this isn't your code).

Reply:
Resolved with commit 1282790529b1bf8cf5af26da431e3746b4be21c3.